### PR TITLE
Initial commit for CUDA fault injection tool

### PR DESCRIPTION
**Pull request by gerashegalov**
This PR is the initial version of CUDA fault injection tool to explore and test for correctness of CUDA error handling in fault-tolerant CUDA applications.

The tool is designed with automated testing and interactive testing use cases in mind. The tool is a dynamically linked library `libcufaultinj.so` that is loaded by the CUDA process via CUDA Driver API `cuInit` if it's provided via the `CUDA_INJECTION64_PATH` environment variable.

As an example it can be used to test RAPIDS Accelerator for Apache Spark. 

### Local Mode
```bash
CUDA_INJECTION64_PATH=$PWD/target/cmake-build/faultinj/libcufaultinj.so \ 
FAULT_INJECTOR_CONFIG_PATH=src/test/cpp/faultinj/test_faultinj.json \
$SPARK_HOME/bin/pyspark \ 
  --jars $SPARK_RAPIDS_REPO/dist/target/rapids-4-spark_2.12-22.08.0-SNAPSHOT-cuda11.jar \ 
  --conf spark.plugins=com.nvidia.spark.SQLPlugin
```
### Distributed Mode
```bash
$SPARK_HOME/bin/spark-shell \
  --jars $SPARK_RAPIDS_REPO/dist/target/rapids-4-spark_2.12-22.08.0-SNAPSHOT-cuda11.jar \
  --conf spark.plugins=com.nvidia.spark.SQLPlugin \
  --files ./target/cmake-build/faultinj/libcufaultinj.so,./src/test/cpp/faultinj/test_faultinj.json \
  --conf spark.executorEnv.CUDA_INJECTION64_PATH=./libcufaultinj.so \
  --conf spark.executorEnv.FAULT_INJECTOR_CONFIG_PATH=test_faultinj.json \
  --conf spark.rapids.memory.gpu.minAllocFraction=0 \
  --conf spark.rapids.memory.gpu.allocFraction=0.2 \
  --master spark://hostname:7077 
```
When we configure the executor environment spark.executorEnv.CUDA_INJECTION64_PATH we have to use a path separator in the value ./libcufaultinj.so with the leading dot to make sure that dlopen loads the library file submitted. Otherwise it will assume a locally installed library accessible to the dynamic linker via LD_LIBRARY_PATH and similar mechanisms. See [dlopen man page](https://man7.org/linux/man-pages/man3/dlopen.3.html)

### Fault injection configuration 

Fault injection configuration is provided via the `FAULT_INJECTOR_CONFIG_PATH` environment variable. It's a set of rules to apply fault injection when CUDA Drvier or Runtime is matched by function name or callback id with a given probability.

There are currently three types of fault injection:
- launch a kernel with the PTX `trap` instruction
- launch a kernel with a device assert
- replace the return code for the CUDA Runtime call

Example config:
```json
{
    "logLevel": 1,
    "dynamic": true,
    "cudaRuntimeFaults": {
        "cudaLaunchKernel_ptsz": {
            "percent": 0,
            "injectionType": 0,
            "injectionType_comment": "PTX trap = 0, C assert = 1",
            "interceptionCount": 1
        }
    },
    "cudaDriverFaults": {
        "cuMemFreeAsync_ptsz": {
            "percent": 0,
            "injectionType": 2,
            "injectionType_comment": "substitute return code",
            "substituteReturnCode": 999,
            "interceptionCount": 1
        }
    }
}
```

Signed-off-by: Gera Shegalov <gera@apache.org>

### CODE REVIEW

File: src/main/cpp/faultinj/faultinj.cu
Comment (jlowe) [inline src/main/cpp/faultinj/faultinj.cu:119]: Curious if this is being pedantic, or is there something that is being cleaned up here that normal OS process teardown will not automatically handle?

File: src/main/cpp/faultinj/faultinj.cu
Comment (jlowe) [inline src/main/cpp/faultinj/faultinj.cu:213]: Is this still relevant?

**Review by jlowe — COMMENTED** [review]
I didn't get into the deep details in the code.  It would be good to have someone with more C++/CUDA experience take a look as well.

I'd like to see the nice writeup in the PR description made into a markdown doc and checked in with this code.  That way there's a user guide sitting next to the code rather than hoping someone does the software archeology to trace the code back to this PR in hopes there's a description of what it does.

Are there plans for packaging this?  It currently builds alongside spark-rapids-jni but doesn't really need it. It's not a tool that requires a Spark/Java application, so wondering if this is the best long-term place for this code.

File: src/main/cpp/faultinj/faultinj.cu
Comment (mythrocks) [inline src/main/cpp/faultinj/faultinj.cu:52]: Not a deal breaker, but I'm curious why we went with a macro, when we had lambdas at our disposal (`CMAKE_CXX_STANDARD 17`).
I'm guessing it's because the call site is simplified if we use a macro?

File: src/main/cpp/faultinj/faultinj.cu
Comment (mythrocks) [inline src/main/cpp/faultinj/faultinj.cu:89]: This is a CUDA file, no? And we have `CMAKE_CXX_STANDARD 17`? Perhaps we could dispense with the `__cplusplus` checks?

**Review by mythrocks — CHANGES_REQUESTED** [review]

File: src/main/cpp/faultinj/faultinj.cu
Comment (mythrocks) [reply src/main/cpp/faultinj/faultinj.cu:117]: Actually, I should read this more closely. I'm not sure if the usage model is:
- tag a separately running Spark RAPIDS process to inject a fault, or,
- inject fault from within the same Spark process.

File: src/main/cpp/faultinj/faultinj.cu
Comment (mythrocks) [inline src/main/cpp/faultinj/faultinj.cu:326]: Nitpick:

I wonder if we might streamline the initialization/management of `globalControl` state from one place. It looks like `globalControl.initialized` is first set to `0`, then changed to `1` *after* (and *outside*) `globalControlInit()`. 

Would there be value in using a `static` constructor to initialize this singleton? It might obviate the need for `initialized`:

```c++
injGlobalControl& STDCALL
￼globalControl(void) {
￼    spdlog::info("InitializeInjection");
     
     static injGlobalControl instance;
     // Guaranteed to be initialized, thread-safely.

     // Move PTRHEAD_CALL(), `cuptiInitialize()`, etc. into body of constructor.

     return &instance;
}
```

`instance` would be initialized with thread-safety, on first call.

Comment (gerashegalov) [reply src/main/cpp/faultinj/faultinj.cu:357]: I think there are many opportunities to make the initial code look more C++ than C. Is it OK if we handle as a separate epic with various suggestions you are making

File: src/main/cpp/faultinj/faultinj.cu
Comment (mythrocks) [inline src/main/cpp/faultinj/faultinj.cu:407]: Thank you for this bit: I had to go look up how we'd do this in C++17: [`std::shared_mutex`](https://en.cppreference.com/w/cpp/thread/shared_mutex), coupled with [`std::unique_lock`](https://en.cppreference.com/w/cpp/thread/unique_lock) for RAII.

Looks like this is doable with the standard library. We can address this later (should we choose to address it at all).

File: src/main/cpp/faultinj/faultinj.cu
Comment (mythrocks) [reply src/main/cpp/faultinj/faultinj.cu:357]: Yes, sir. Of course. None of what I'm pointing out warrants blocking this.

File: src/main/cpp/faultinj/faultinj.cu
Comment (mythrocks) [reply src/main/cpp/faultinj/faultinj.cu:48]: More than anything else, this is a reflection of my prejudice against macros. Your code looks correct, at least to my eye.

File: src/main/cpp/faultinj/faultinj.cu
Comment (mythrocks) [inline src/main/cpp/faultinj/faultinj.cu:469]: TIL about `inotify`, I'm ashamed to admit.

File: src/main/cpp/faultinj/faultinj.cu
Comment (mythrocks) [reply src/main/cpp/faultinj/faultinj.cu:114]: > dynamically linked to the process under test

Ah, then @jlowe might have a point about CUPTI cleanup likely happening on process exit.
That said, my vote would be to keep the `atExitHandler()` anyway.

File: src/main/cpp/faultinj/README.md
Comment (jlowe) [inline src/main/cpp/faultinj/README.md:20]: This should be wrapped to a reasonable length like the other lines for viewing in a terminal

File: src/main/cpp/faultinj/README.md
Comment (jlowe) [inline src/main/cpp/faultinj/README.md:22]:
```suggestion
As an example it can be used to test RAPIDS Accelerator for Apache Spark. Consult documentation to find
```

File: src/main/cpp/faultinj/README.md
Comment (mythrocks) [inline src/main/cpp/faultinj/README.md:8]:
```suggestion
- fatal errors leaving a GPU in unusable state are detected and such GPUs are prevented from executing retries for
```

File: src/main/cpp/faultinj/README.md
Comment (mythrocks) [inline src/main/cpp/faultinj/README.md:15]:
```suggestion
or the higher-level framework such as Apache Spark remains usable when it should,
```

File: src/main/cpp/faultinj/README.md
Comment (mythrocks) [inline src/main/cpp/faultinj/README.md:47]: Should the code bits be quoted?
```suggestion
When we configure the executor environment `spark.executorEnv.CUDA_INJECTION64_PATH`
```

File: src/main/cpp/faultinj/README.md
Comment (mythrocks) [inline src/main/cpp/faultinj/README.md:3]: Brilliant work here. Thank you for documenting this.

**Review by mythrocks — DISMISSED** [review]
I'm largely :+1: on this. There are some minor fixes pending in documentation.

There are parts of this we should probably revisit in the future:
- RAII for (RW)locks, closing files, etc.
- Maybe manage singleton state differently, for `globalConfig`.
- Maybe switch out of using `pthreads` directly, in favour of C++ standard threads.

File: src/main/cpp/faultinj/README.md
Comment (mythrocks) [inline src/main/cpp/faultinj/README.md:6]: Super pedantic nitpick: I would avoid this sort of contraction in documentation, if possible:
```suggestion
It is especially important when CUDA is embedded in a higher-level fault-tolerant framework
```

File: src/main/cpp/faultinj/README.md
Comment (gerashegalov) [reply src/main/cpp/faultinj/README.md:3]: thank you!

File: src/main/cpp/faultinj/faultinj.cu
Comment (mythrocks) [inline src/main/cpp/faultinj/faultinj.cu:323]: Apologies for not catching this sooner: I think we might need a `cudaStreamSynchronize(0)` or something here. It's a tiny kernel.

**Review by mythrocks — CHANGES_REQUESTED** [review]
I think we need a stream sync right after the kernel launches, given how short the kernel are. Otherwise, we run the risk of exiting before the kernel has launched.

File: src/main/cpp/CMakeLists.txt
Comment (ttnghia) [inline src/main/cpp/CMakeLists.txt:257]: Maybe everything called `fault_injection` is more expressive and better than `faultinj`.

File: src/main/cpp/faultinj/README.md
Comment (ttnghia) [inline src/main/cpp/faultinj/README.md:11]:
```suggestion
framework to ensure that CUDA failures are handled correctly in that:
 * Fatal errors leaving a GPU in unusable state are detected and such GPUs are
prevented from executing retries for a failed computation
 * Non-fatal errors are retried without losing valuable compute resources
 * Error handling logic does not cause deadlocks and other sort of unresponsiveness.
```

File: src/main/cpp/faultinj/README.md
Comment (ttnghia) [inline src/main/cpp/faultinj/README.md:69]: Similar to above.

File: src/main/cpp/faultinj/faultinj.cu
Comment (ttnghia) [reply src/main/cpp/faultinj/faultinj.cu:48]: If you want to print the crash line using `__LINE__` native macro then a new macro is necessary. Otherwise either macro or function will work. But a function is preferred.

File: src/main/cpp/faultinj/faultinj.cu
Comment (ttnghia) [inline src/main/cpp/faultinj/faultinj.cu:88]: Global variables are discouraged as they may have some potential side effects with their initialization order.

File: src/main/cpp/faultinj/faultinj.cu
Comment (ttnghia) [inline src/main/cpp/faultinj/faultinj.cu:56]: To prevent name clash, all the symbols, except the one that will be called outside of this file, should be wrapped in anonymous namespace:
```
namespace {
...
}
```

File: src/main/cpp/faultinj/faultinj.cu
Comment (ttnghia) [inline src/main/cpp/faultinj/faultinj.cu:79]: Functions without attaching to any class can be just functions, don't need to have `static` keyword as they are already static.

File: src/main/cpp/faultinj/faultinj.cu
Comment (ttnghia) [inline src/main/cpp/faultinj/faultinj.cu:475]: Please group `constexpr` and other variables in different groups, don't mix them.

File: src/main/cpp/faultinj/faultinj.cu
Comment (ttnghia) [inline src/main/cpp/faultinj/faultinj.cu:505]: In C++, `NULL` is discouraged. Use `nullptr` for pointer instead.

File: src/main/cpp/faultinj/faultinj.cu
Comment (gerashegalov) [reply src/main/cpp/faultinj/faultinj.cu:36]: at the very least we need to export InitializeInjection for cuInit. There may be currently also unnecessary symbols in this scope

File: src/main/cpp/faultinj/faultinj.cu
Comment (gerashegalov) [reply src/main/cpp/faultinj/faultinj.cu:323]: I had a bunch of iterations here https://github.com/NVIDIA/spark-rapids-jni/pull/399/commits/36d2678c590d001f614a42d251b26ce159a7dfcf . Let us do cudaStreamSynchronize then :)

File: src/main/cpp/CMakeLists.txt
Comment (gerashegalov) [reply src/main/cpp/CMakeLists.txt:257]: We can reconsider, it feels too long for a directory name at the moment. It contains a README.md with enough background

File: src/main/cpp/faultinj/faultinj.cu
Comment (gerashegalov) [reply src/main/cpp/faultinj/faultinj.cu:79]: after we start using anonymous namespace which I'll try

**Comment by gerashegalov** [comment]
build

**Review by mythrocks — DISMISSED** [review]
I'm :+1:. Thank you for accommodating the changes requested.

There might be future iterations where we might make C++-related changes. E.g. RIAA for mutex locking, `std::mutex` and threads instead of `pthreads` primitives. But we needn't bother with those now.

**Comment by mythrocks** [comment]
> [mythrocks](https://github.com/mythrocks) approved these changes [9 minutes ago](https://github.com/NVIDIA/spark-rapids-jni/pull/399#pullrequestreview-1055874081)
...
> Review required
At least 1 approving review is required by reviewers with write access.

Notice that my :+1: is ignored here. :]

File: src/main/cpp/faultinj/faultinj.cu
Comment (ttnghia) [inline src/main/cpp/faultinj/faultinj.cu:90]: Style needs to be consistent. As a more widely used style: please do not break line between return type and function name.

Comment (ttnghia) [reply src/main/cpp/faultinj/faultinj.cu:90]: I wonder if you use any formatter? We should use clang-format.

File: src/main/cpp/faultinj/faultinj.cu
Comment (ttnghia) [inline src/main/cpp/faultinj/faultinj.cu:226]: Should be `else if`.

File: src/main/cpp/faultinj/faultinj.cu
Comment (ttnghia) [inline src/main/cpp/faultinj/faultinj.cu:341]:
```suggestion
    case FI_RETURN_VALUE:
        if (domain == CUPTI_CB_DOMAIN_DRIVER_API) {
            CUresult *cuResPtr = static_cast<CUresult*>(cbInfo->functionReturnValue);
            *cuResPtr = static_cast<CUresult>(substituteReturnCode);
        } else if (domain == CUPTI_CB_DOMAIN_RUNTIME_API) {
            cudaError_t *cudaErrPtr = static_cast<cudaError_t*>(cbInfo->functionReturnValue);
            spdlog::error("updating runtime return value DOES NOT WORK, use trap or assert");
            *cudaErrPtr = static_cast<cudaError_t>(substituteReturnCode);
        }
        break;
```

File: src/main/cpp/faultinj/faultinj.cu
Comment (ttnghia) [inline src/main/cpp/faultinj/faultinj.cu:344]:
```suggestion
    default: ;
```

File: src/main/cpp/faultinj/faultinj.cu
Comment (ttnghia) [inline src/main/cpp/faultinj/faultinj.cu:416]: Don't have to cache `end`.

```
for(auto it = pTree.begin(); it != pTreen.end(); ++it)
```

File: src/main/cpp/faultinj/faultinj.cu
Comment (gerashegalov) [reply src/main/cpp/faultinj/faultinj.cu:90]: We need pre-commit hool like cudf to do that like in cudf.

Comment (ttnghia) [reply src/main/cpp/faultinj/faultinj.cu:90]: I would recommend to use the clang-format config from cudf. Our IDE can just load the config and do its job.

Comment (gerashegalov) [reply src/main/cpp/faultinj/faultinj.cu]: any style works for me, especially if it is just done automatically. Tieing it to the editor is ok but not ideal.

File: src/main/cpp/faultinj/faultinj.cu
Comment (gerashegalov) [reply src/main/cpp/faultinj/faultinj.cu]: I find it more readable the way it's symmetrically written, but either way is fine with me in this case.

File: src/main/cpp/faultinj/faultinj.cu
Comment (gerashegalov) [reply src/main/cpp/faultinj/faultinj.cu]: any style works for me, especially if it is just done automatically. Tieing it to the editor is ok but not ideal.

**Comment by gerashegalov** [comment]
> There might be future iterations where we might make C++-related changes. E.g. RIAA for mutex locking, `std::mutex` and threads instead of `pthreads` primitives. But we needn't bother with those now.

Thanks for the review @mythrocks 

absolutely, once the PR is merged I'll file an epic for these improvements

**Comment by gerashegalov** [comment]
build

File: src/main/cpp/faultinj/faultinj.cu
Comment (ttnghia) [inline src/main/cpp/faultinj/faultinj.cu:337]:
```suggestion
      }
```

**Review by ttnghia — APPROVED** [review]

**Review by jlowe — APPROVED** [review]

**Pull request merged**
